pkg/util/annotation.go

```diff
@@ -29,3 +47,38 @@ func GetAnnotationValue(annotations map[string]string, annotationKey string) str
 	}
 	return annotations[annotationKey]
 }
+
+func getDeletedAnnotationKeys(desired, observed *unstructured.Unstructured) sets.Set[string] {
+	recordKeys := sets.NewString(strings.Split(observed.GetAnnotations()[workv1alpha2.ManagedAnnotation], ",")...)
+	for key := range desired.GetAnnotations() {
+		recordKeys.Delete(key)
+	}
+	return sets.Set[string](recordKeys)
+}
+
+// RecordLabelAndAnnotationKey record labels and annotations key.
+func RecordLabelAndAnnotationKey(object *unstructured.Unstructured) {
```

Comment:
Can we do it with two separated functions? like:
```golang
// RecordManagedAnnotations sets or updates the annotation(resourcetemplate.karmada.io/managed-annotations)
// to record the annotation keys.
// Note: Karmada defined keys(contains "karmada.io") will be ignored.
func RecordManagedAnnotations(object *unstructured.Unstructured) {
	annotations := object.GetAnnotations()
	var managedKeys []string

	for key, _ := range annotations {
		// ignore the keys that defined by Karmada
		if strings.Contains(key, "karmada.io") {
			continue
		}

		managedKeys = append(managedKeys, key)
	}

	if len(managedKeys) > 0 {
		sort.Strings(managedKeys)
		annotations[workv1alpha2.ManagedAnnotation] = strings.Join(managedKeys, ",")
		object.SetAnnotations(annotations)
	}
}
```
The other one could be `func RecordManagedLabels(object *unstructured.Unstructured)`.

In addition, I think the keys that defined by Karmada should be ignored.

Reply:
> In addition, I think the keys that defined by Karmada should be ignored.

Wait, I just realized if we ignore the Karmada keys, then we probably don't have a chance to remove it even it is deprecated, right?

Reply:
Yes, I think only the newly added two annotations can be filtered, and this description may need to be modified:
`the labels defined by Karmada will be ignored to avoid clutter`

Reply:
> Wait, I just realized if we ignore the Karmada keys, then we probably don't have a chance to remove it even it is deprecated, right?

Yes, so we must record all keys. Whether it's Karmada or others.

Reply:
>  and this description may need to be modified:
the labels defined by Karmada will be ignored to avoid clutter

What do you suggest?

Reply:
For the above reasons, we may not be able to ignore karmada keys, so this comment may no longer be appropriate and in my opinion needs to be deleted, @RainbowMango 's confirmation is required.

Reply:
Yes. the comment on the annotation should be removed:
> the labels defined by Karmada will be ignored to avoid clutter and

Reply:
https://github.com/karmada-io/karmada/pull/3099#discussion_r1099782042

``` go
      if len(managedKeys) > 0 {
		sort.Strings(managedKeys)
		annotations[workv1alpha2.ManagedAnnotation] = strings.Join(managedKeys, ",")
		object.SetAnnotations(annotations)
	}
```


I think if the `managedKeys` is empty we also record it to `workv1alpha2.ManagedAnnotation`.

Reply:
Why do you think that? What's the benefit of having an empty key set?

Reply:
Users can clearly know that no key is currently managed by Karmada in the member cluster.

Reply:
Yeah, It makes sense.
Then, let's do it.